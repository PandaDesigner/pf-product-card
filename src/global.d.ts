// global.d.ts
declare module '*.module.css' {
    const classes: { [ key: string ]: string };
    export default classes;
}

declare module '*.css' {
    const classes: { [ className: string ]: string };
    export default classes;
}

declare module '*.jpg' {
    const value: any;
    export default value;
}
