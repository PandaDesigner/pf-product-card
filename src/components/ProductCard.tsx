import { CSSProperties, createContext } from 'react';
import { useProduct } from "../hooks/useProduct";
// @ts-ignore
import styles from '../styles/styles.module.css';

import {
  InitialValues,
  OnChangeArgs,
  Product,
  ProductCardHandlers,
  ProductContextProps,
} from "../interfaces/interfaces";
import React from 'react';

export const ProductContext = createContext({} as ProductContextProps);

const { Provider } = ProductContext;

export interface ProductCardProps {
  product: Product;
  children?: (arg: ProductCardHandlers) => JSX.Element;
  className?: string;
  style?: CSSProperties;
  onChange?: (args: OnChangeArgs) => void;
  value?: number;
  initialValues?: InitialValues;
}

export const ProductCard = ({
  children,
  product,
  className,
  style,
  onChange,
  value,
  initialValues,
}: ProductCardProps) => {
  const { counter, increaseBy, maxCount, isMaxCountReached, reset } =
    useProduct({
      onChange,
      product,
      value,
      initialValues,
    });

  return (
    <Provider
      value={{
        counter,
        increaseBy,
        product,
        maxCount,
        reset,
      }}
    >
      <div className={`${styles.productCard} ${className}`} style={style}>
        {children
          ? children({
            count: counter,
            isMaxCountReached,
            maxCount: initialValues?.maxCount,
            product,
            increaseBy,
            reset,
          })
          : null}
      </div>
    </Provider>
  );
};
