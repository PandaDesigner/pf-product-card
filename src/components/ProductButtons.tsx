import { CSSProperties, useCallback, useContext } from 'react';
import { ProductContext } from './ProductCard';
// @ts-ignore
import styles from '../styles/styles.module.css';
import React from 'react';

export interface PropsButtons {
  className?: string;
  style?: CSSProperties;
}

export const ProductButtons = ({ className, style }: PropsButtons) => {
  const { counter, increaseBy, maxCount } = useContext(ProductContext);

  const isMaxReached = useCallback(
    () => !!maxCount && counter === maxCount,
    [counter, maxCount]
  );

  const isMinReached = useCallback(() => counter < 1, [counter]);

  return (
    <div className={`${styles.buttonsContainer} ${className}`} style={style}>
      <button
        className={`${styles.buttonMinus} ${isMinReached() && styles.disable__right
          }`}
        onClick={() => increaseBy(-1)}
        disabled={isMinReached()}
      >
        -
      </button>
      <div className={styles.countLabel}> {counter} </div>
      <button
        className={`${styles.buttonAdd} ${isMaxReached() && styles.disable__left
          }`}
        onClick={() => increaseBy(1)}
        disabled={isMaxReached()}
      >
        +
      </button>
    </div>
  );
};
