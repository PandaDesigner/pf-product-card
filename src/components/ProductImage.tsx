/* eslint-disable jsx-a11y/img-redundant-alt */
import { CSSProperties, useContext } from 'react';
import { ProductContext } from './ProductCard';
// @ts-ignore
import noImage from '../assets/no-image.jpg';
// @ts-ignore
import styles from '../styles/styles.module.css';
import React from 'react';

export interface PropsImage {
  imgs?: string;
  className?: string;
  style?: CSSProperties;
}

export const ProductImage = ({ imgs = '', className, style }: PropsImage) => {
  const { product } = useContext(ProductContext);

  let imgToShow: string;

  if (imgs) {
    imgToShow = imgs;
  } else if (product.img) {
    imgToShow = product.img;
  } else {
    imgToShow = noImage;
  }

  // @ts-ignore
  return (
    <img
      className={`${styles.productImg} ${className}`}
      src={imgToShow}
      alt='Prooduct Image'
      style={style}
    />
  );
};
