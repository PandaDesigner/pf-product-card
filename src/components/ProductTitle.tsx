import { CSSProperties, useContext } from 'react';
// @ts-ignore
import styles from '../styles/styles.module.css';
import { ProductContext } from './ProductCard';
import React from 'react';

export interface PropsTitle {
  title?: string;
  className?: string;
  style?: CSSProperties;
}

export const ProductTitle = ({ title, className, style }: PropsTitle) => {
  const { product } = useContext(ProductContext);

  return (
    <span className={`${styles.productDescription} ${className}`} style={style}>
      {title ? title : product.title}
    </span>
  );
};
