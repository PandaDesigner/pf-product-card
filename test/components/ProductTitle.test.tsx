import React from 'react';
import renderer from 'react-test-renderer'
import { ProductTitle, ProductCard } from '../../src/components';
import { PRODUCT } from './data/products';

describe('ProductTitle', () => {
    test('should display the component correctly with the custom title', () => {
        const wrapper = renderer.create(
            <ProductTitle title='custom product' className='custom' />
        )
        expect(wrapper.toJSON()).toMatchSnapshot();
    });

    test('should display the component with the product name', () => {

        const wrapper = renderer.create(
            <ProductCard product={PRODUCT.withoutImage}>
                {() => <ProductCard.Title className="productDescription" />}
            </ProductCard>
        );
        //@ts-expect-error
        const titleNode = wrapper.toJSON()?.children?.[0]?.children?.[0] || null;

        expect(titleNode).toBe(PRODUCT.withoutImage.title)
        expect(wrapper.toJSON()).toMatchSnapshot();
    });
})