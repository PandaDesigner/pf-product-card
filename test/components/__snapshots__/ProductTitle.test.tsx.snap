// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductTitle should display the component correctly with the custom title 1`] = `
<span
  className="productDescription custom"
>
  custom product
</span>
`;

exports[`ProductTitle should display the component with the product name 1`] = `
<div
  className="productCard undefined"
>
  <span
    className="productDescription productDescription"
  >
    Coffee Mug - Card
  </span>
</div>
`;
